crypto/keyring_streaming.go

```diff
@@ -0,0 +1,275 @@
+// and returns a writer for the plaintext data and the key packet.
+// If signKeyRing is not nil, it is used to do an embedded signature.
+func (keyRing *KeyRing) EncryptSplitStream(
+	dataPacketWriter Writer,
+	isBinary bool,
+	filename string,
+	modTime int64,
+	signKeyRing *KeyRing,
+) (*EncryptSplitResult, error) {
+	config := &packet.Config{DefaultCipher: packet.CipherAES256, Time: getTimeGenerator()}
+
+	hints := &openpgp.FileHints{
+		IsBinary: isBinary,
+		FileName: filename,
+		ModTime:  time.Unix(modTime, 0),
+	}
+	var keyPacketBuf bytes.Buffer
+	plainMessageWriter, err := asymmetricEncryptStream(hints, &keyPacketBuf, dataPacketWriter, keyRing, signKeyRing, config)
+	if err != nil {
+		return nil, err
+	}
+	return &EncryptSplitResult{
+		keyPacketBuf:       &keyPacketBuf,
+		plainMessageWriter: plainMessageWriter,
+	}, nil
+}
+
+// PlainMessageReader is used to wrap the data of the decrypted plain message.
+// It can be used to read the decrypted data and verify the embedded signature.
+type PlainMessageReader struct {
+	details       *openpgp.MessageDetails
+	verifyKeyRing *KeyRing
+	verifyTime    int64
+	readAll       bool
+}
+
+// IsBinary returns whether the message is binary or text.
+func (msg *PlainMessageReader) IsBinary() bool {
+	return msg.details.LiteralData.IsBinary
+}
+
+// GetFilename returns the filename of the message.
+func (msg *PlainMessageReader) GetFilename() string {
+	return msg.details.LiteralData.FileName
+}
+
+// GetModificationTime returns the modification time of the message.
+func (msg *PlainMessageReader) GetModificationTime() int64 {
+	return int64(msg.details.LiteralData.Time)
+}
+
+// Read is used to access the message decrypted data.
+// Makes PlainMessageReader implement the Reader interface.
+func (msg *PlainMessageReader) Read(b []byte) (n int, err error) {
+	n, err = msg.details.UnverifiedBody.Read(b)
+	if errors.Is(err, io.EOF) {
+		msg.readAll = true
+	}
+	return
+}
+
+// VerifySignature is used to verify that the signature is valid.
+// This method needs to be called once all the data has been read.
+// It will return an error if the signature is invalid
+// or if the message hasn't been read entirely.
+func (msg *PlainMessageReader) VerifySignature() (err error) {
+	if !msg.readAll {
+		return errors.New("gopenpg: can't verify the signature until the message reader has been read entirely")
+	}
+	if msg.verifyKeyRing != nil {
+		processSignatureExpiration(msg.details, msg.verifyTime)
+		err = verifyDetailsSignature(msg.details, msg.verifyKeyRing)
+	}
+	return
+}
+
+// DecryptStream is used to decrypt a pgp message as a Reader.
+// It takes a reader for the message data
+// and returns a PlainMessageReader for the plaintext data.
+// If verifyKeyRing is not nil, PlainMessageReader.VerifySignature() will
+// verify the embedded signature with the given key ring and verification time.
+func (keyRing *KeyRing) DecryptStream(
+	message Reader,
+	verifyKeyRing *KeyRing,
+	verifyTime int64,
```

Comment:
For the verify time we could, but for the verification key ring, go-crypto/openpgp needs it before starting the decryption. At least I haven't found a way to give it afterward. 
So I kept them both at the start, because it keeps consistency of having both verify keyring and verify time together.

Reply:
Let's keep it consistent :+1: